src/DI/ContainerBuilder.php

```diff
@@ -783,59 +582,64 @@ public function formatStatement(Statement $statement)
 			if ($class) {
 				$arguments = $this->autowireArguments($class, $entity[1], $arguments);
 			}
-			return $this->formatPhp('?->?(?*)', [$entity[0], $entity[1], $arguments]);
+			$entity[0] = '@' . $service;
 
 		} else { // static method
 			$arguments = $this->autowireArguments($entity[0], $entity[1], $arguments);
-			return $this->formatPhp("$entity[0]::$entity[1](?*)", [$arguments]);
 		}
-	}
-
 
-	/**
-	 * Formats PHP statement.
-	 * @return string
-	 * @internal
-	 */
-	public function formatPhp($statement, $args)
-	{
-		array_walk_recursive($args, function (& $val) {
+		array_walk_recursive($arguments, function (& $val) {
 			if ($val instanceof Statement) {
-				$val = self::literal($this->formatStatement($val));
-
-			} elseif ($val === $this) {
-				$val = self::literal('$this');
+				$val = $this->completeStatement($val);
 
 			} elseif ($val instanceof ServiceDefinition) {
 				$val = '@' . current(array_keys($this->getDefinitions(), $val, TRUE));
-			}
-
-			if (!is_string($val)) {
-				return;
 
-			} elseif (substr($val, 0, 2) === '@@') {
-				$val = substr($val, 1);
-
-			} elseif (substr($val, 0, 1) === '@' && strlen($val) > 1) {
+			} elseif (is_string($val) && strlen($val) > 1 && $val[0] === '@' && $val[1] !== '@') {
 				$pair = explode('::', $val, 2);
 				$name = $this->getServiceName($pair[0]);
-				if (isset($pair[1]) && preg_match('#^[A-Z][A-Z0-9_]*\z#', $pair[1])) {
-					$val = $this->getDefinition($name)->getClass() . '::' . $pair[1];
-				} elseif (isset($pair[1])) {
-					$val = $this->formatStatement(new Statement(['@' . $name, '$' . $pair[1]]));
-				} else {
-					if ($name === self::THIS_CONTAINER) {
-						$val = '$this';
-					} elseif ($name === $this->currentService) {
-						$val = '$service';
-					} else {
-						$val = $this->formatStatement(new Statement(['@' . self::THIS_CONTAINER, 'getService'], [$name]));
-					}
+				if (!isset($pair[1])) { // @service
+					$val = '@' . $name;
+				} elseif (preg_match('#^[A-Z][A-Z0-9_]*\z#', $pair[1], $m)) { // @service::CONSTANT
+					$val = self::literal($this->getDefinition($name)->getClass() . '::' . $pair[1]);
+				} else { // @service::property
+					$val = new Statement(['@' . $name, '$' . $pair[1]]);
 				}
-				$val = self::literal($val);
 			}
 		});
-		return PhpHelpers::formatArgs($statement, $args);
+
+		return new Statement($entity, $arguments);
+	}
+
+
+	private function checkCase($class)
+	{
+		if ((class_exists($class) || interface_exists($class)) && $class !== ($name = (new ReflectionClass($class))->getName())) {
+			throw new ServiceCreationException("Case mismatch on class name '$class', correct name is '$name'.");
+		}
+	}
+
+
+	/**
+	 * Adds a file to the list of dependencies.
+	 * @return self
+	 * @internal
+	 */
+	public function addDependency($file)
```

Comment:
This is actually a very usefull method and IMHO it shouldn't be marked internal.